app/src/main/java/org/thoughtcrime/securesms/util/FlowUtils.kt

```diff
@@ -0,0 +1,63 @@
+package org.thoughtcrime.securesms.util
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.flatMapConcat
+
+/**
+ * Buffers items from the flow and emits them in batches. The batch will have size [maxItems] and
+ * time [timeoutMillis] limit.
+ */
+fun <T> Flow<T>.timedBuffer(timeoutMillis: Long, maxItems: Int): Flow<List<T>> {
+    class State(
+        var buffer: MutableList<T> = mutableListOf(),
+        var deadline: Long = -1L
+    )
+
+    return scanTransformLatest(State()) { state, value ->
+        val now = System.currentTimeMillis()
+
+        if (state.buffer.isEmpty()) {
+            state.deadline = now + timeoutMillis
+        }
+
+        state.buffer.add(value)
+
+        if (state.buffer.size < maxItems && now < state.deadline) {
+            // If the buffer is not full and the timeout has not expired, keep waiting
+            // until the deadline is reached. This delay will get cancelled and our
+            // transform function will be called again for the new value.
+            delay(state.deadline - now)
+        }
+
+        // When we reach here, the buffer is either full or the timeout has expired,
+        // at which point we emit our buffer.
+        val result = state.buffer
+        state.buffer = mutableListOf()
+        state.deadline = -1L
+
+        result
+    }
+}
+
+/**
+ * Scans the flow with a state and transform the latest value with the state.
+ * This function is similar to [Flow.scan] but it differs in a major ways:
+ * 1. The accumulator used by this function returns the transformed value rather than the state.
+ *    This then potentially requires the state to be mutable.
+ * 2. The transform function will get cancelled for the new value (this is where the name "latest" comes from).
+ */
+fun <T, S, R> Flow<T>.scanTransformLatest(state: S, transform: suspend (acc: S, value: T) -> R): Flow<R> {
```

Comment:
not very "functional" code, requires mutation to do anything interesting, you don't even need `state` here, as it never changes. We always call `transform` with the initial value.

In the usage, it looks like the code will do something else. It looks like state can/will change, just because of the idiom.

Reply:
Mutation is not the end of the world, this particular one saves a lot of trouble and a lot of interim object creation. I believe as long as you can guarantee the state is only access by one suspend function at any time, the mutability is not a concern.

I get this idea from Rust's iterator's `scan`: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.scan.

Reply:
Your function is a bit different.

They have what, a reference, and the reference actually changes as they scan through, they're using immutable type int. This referenence to a State we have here does not change between calls. So omitting state from the function achieves the same and is simpler to reason about:

```
fun <T> Flow<T>.timedBuffer(timeoutMillis: Long, maxItems: Int): Flow<List<T>> {
    class State(
        var buffer: MutableList<T> = mutableListOf(),
        var deadline: Long = -1L
    )
    
    val state = State()

    return scanTransformLatest { value ->
        val now = System.currentTimeMillis()

        if (state.buffer.isEmpty()) {
            state.deadline = now + timeoutMillis
        }
...
```

Alternatively the State could change between calls using some `Ref` or similar under the hood, but atm this implementation is confusing because `State` param is not really necessary, and the implication is that state will change in different calls.

Reply:
Ah I see what you mean. I want `scanTransformLatest` to take the state object, is because:
1. It aligns with what "scan" means, it requires a state to function, albeit a mutable one. The function signature is significant in this sense.
2. To use a state outside like what you suggested will look to me as further deviation from functional style, at least with my original approach, the state is still contained inside the flow.

I don't think we are gonna get a resolution here so let me just change to another conservative approach.

Reply:
you can either use a `Ref` in your impl, or just use `.map {` no?

Reply:
@bemusementpark @ThomasSession wanna have a look again?
So the problem is the introduction of mutability is always going to be considered imperfect, so I've removed the `scanTransformLatest` and just do a simpler version of using `collectLatest` in the implementation of `timedBuffer`, hope it makes sense